File: src/main/scala/operator/FlatMapOperator.scala
package main.scala.operator

import org.apache.spark.{SparkContext, SparkConf}

/**
  * Created by root on 2016/6/7.
  */
object FlatMapOperator {

  def main(args: Array[String]) {
    val conf = new SparkConf().setAppName("FlatMapOperator").setMaster("local")
    val sc = new SparkContext(conf)

    val lineArray = Array("hello xuruyun" , "hello xuruyun", "hello wangfei")
    val linesRDD = sc.parallelize(lineArray)
    // {hello,xuruyun,hello,xuruyun,hello,wangfei}
    
    /**
     *  hello
        xuruyun
        hello
        xuruyun
        hello
        wangfei
     */
    val wordsWithFlatMap = linesRDD.flatMap(_.split(" "))
    
    
    /**
     * [Ljava.lang.String;@7134ead0
      [Ljava.lang.String;@492437c0
      [Ljava.lang.String;@68c843c3
     */
     val wordsWithMap = linesRDD.map (_.split(" "))
//    
//    
     wordsWithFlatMap.foreach(println _)
    
    
    
//     wordsWithMap.foreach(println _)
    
    
   /* val wordsWithMapResult = wordsWithMap.collect()
    for(a<- wordsWithMapResult){
      println(a.length)
    }*/
  }
}

File: src/main/java/operator/FlatMapOperator.java
package main.java.operator;

import java.util.Arrays;
import java.util.List;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.VoidFunction;

public class FlatMapOperator {

	public static void main(String[] args) {
		SparkConf conf = new SparkConf().setAppName("LineCount").setMaster(
				"local");
		JavaSparkContext sc = new JavaSparkContext(conf);
		
		List<String> lineList = Arrays.asList("hello xuruyun" , "hello xuruyun", "hello wangfei");
		JavaRDD<String> lines = sc.parallelize(lineList);
		
		// flatMap = flat + map
		JavaRDD<String> words = lines.flatMap(new FlatMapFunction<String, String>() {

			private static final long serialVersionUID = 1L;

			@Override
			public Iterable<String> call(String line) throws Exception {
				return Arrays.asList(line.split(" "));
			}
		});
		
		words.foreach(new VoidFunction<String>() {

			private static final long serialVersionUID = 1L;

			@Override
			public void call(String result) throws Exception {
				System.out.println(result);
			}
		});
		
		sc.close();
	}
}
